### PR TITLE
feat: Tracing without Performance

### PR DESCRIPTION
**Pull request by cleptric**
## TL;DR

These changes will kick off a movement where the performance feature is no longer required to implement distributed tracing.

At its core is a new `PropagationContext`, an object holding all required properties to implement distributed tracing in Sentry: a trace ID, span ID, parent span ID, and a Dynamic Sampling Context.

The `PropagationContext` is attached to each scope and is used as a fallback if tracing is not enabled (`Options::isTracingEnabled()` is `false`).

## New API

**Naming is not final!**

### `Sentry\continueTrace`

A new top-level function to continue a trace. This function is context-aware, as in it will either mutate the PropagationContext on the scope or return a populated `TransactionContext` if tracing is enabled, based on the values of a passed in `sentry-trace` and `baggage` header.
While this is not a great API, we have to maintain BC and are willing to live with this for now.

### `Sentry\traceparent`

A new top-level function to receive the trace context to be used as an HTML meta tag for SSR applications or as a header value to be attached to outgoing HTTP requests. This function is context-aware, as in it will either return the trace context of the current span if tracing is enabled or fall back to this value from the PropagationContext stored on the scope.

### `Sentry\baggage`

A new top-level function to receive the Dynamic Sampling Context to be used as an HTML meta tag for SSR applications or as a header value to be attached to outgoing HTTP requests. This function is context-aware, as in it will either return the Dynamic Sampling Context of the current transaction if tracing is enabled or fall back to this value from the PropagationContext stored on the scope.

## Each event always contains a trace context

If tracing is disabled, we will now use the PropagationContext as a fallback to attach a `trace` context to each event, in the case of this SDK, to all error, transaction, and check-in events.

## Each event always contains a Dynamic Sampling Context (trace) envelope header item

As this SDK still uses the `/store` endpoint if tracing is disabled, we can only attach the DSC to check-in events in this case.

## `DynamicSamplingContext::fromOptions`

To propagate a Dynamic Sampling Context in case tracing is disabled, a new method was added that constructs the Dynamic Sampling Context from the current options of the client. 

Part of #1525

### CODE REVIEW

**Review by stayallive — APPROVED** [review]
I have a hard time finding any wrong coce-wise. Looks good!

Just a single TODO we (might) need to address but otherwise good to go!

File: src/Tracing/GuzzleTracingMiddleware.php [hunk 1]
```diff
@@ -27,6 +29,11 @@ public static function trace(?HubInterface $hub = null): Closure
                 $span = $hub->getSpan();
 
                 if (null === $span) {
+                    // TODO(michi) Decide on trace_propagation_targets handling
```
[the remark this answers is not in the document]
Comment (cleptric) [reply src/Tracing/GuzzleTracingMiddleware.php:32]: Yep. So...

We do want to attach both headers by default, always. The current default is `[]`, which we could use as the "always attach" flag.
We could add `null` as an indication to never attach the headers. But this feels odd.

Ideas? 🙂

[the remark this answers is not in the document]
Comment (stayallive) [reply src/Tracing/GuzzleTracingMiddleware.php:32]: Yeah, `['*']` makes more sense as a set always default. But that boat has sailed so maybe `null` is the best bad option and next major we change it to something a little more sensible?

[the remark this answers is not in the document]
Comment (cleptric) [reply src/Tracing/GuzzleTracingMiddleware.php:32]: Agreed, let's roll with this! Will PR as a follow-up.
